### DISCUSSION TITLE
Setting Metals Java Home will result in updating Bloop's global Json file, by Metals

### DISCUSSION THREAD

**Discussion by antosha417** [scalameta/nvim-metals#411] [Q&A] [answered]
Hi! I just updated metals via `Update Metals` command in telescope. And now every time I open scala project I see an annoying message: 
```Setting Metals Java Home will result in updating Bloop's global Json file, by Metals.
Bloop will need to be restarted in order for these changes to take effect.

Request Actions:
1. Apply and restart Bloop
2. Not now
Type number and <Enter> or click with the mouse (q or empty cancels):
```

I do not have bloop installed globally. I believe metals does it by itself and later on asks permission to override configuration. 

After importing a build I see that `~/.bloop/` is created with a two files:
* `bloop.json` 
```json
{"javaHome":"/Users/antosha417/.sdkman/candidates/java/11.0.14-zulu/zulu-11.jdk/Contents/Home"}
```
* `created_by_metals.lock`
```
1654518390504
```

My `$JAVA_HOME` is set to `/Users/antosha417/.sdkman/candidates/java/current`.

**Is there any way that I can disable this notification?**

**Comment by ckipp01** [comment, marked as answer]
Hey @antosha417, you're correct, Metals controls the version of Bloop that it needs and starts it if needed. Bloop however runs as a background process, and Metals doesn't ultimately know if it started it, or if something else started it. So in the situation you're sharing I'm assuming even with the new `bloop.json` etc the old Bloop instance is still running (since it was never told to stop). If you answer `1.` to the prompt to stop and restart Bloop with the new version that Metals needs. Then you shouldn't see it again. Can you try that?

**Reply by antosha417** [reply]
If I select `1.` then I do not see the message anymore. But If I close and reopen the project this message persists. So I see this message every time I open any scala project.

**Reply by ckipp01** [reply]
> But If I close and reopen the project this message persists. So I see this message every time I open any scala project.

Sure, but that's sort of the point. Every time you close and re-open Metals will re-start, realize an old Bloop is running and the version of Metals you're using needs the newer version. So it's intentional that it keep showing you the message in order to restart Bloop. Or are you saying after you choose `1.`  bloop restarts, then you open it again it shows you it again?

**Reply by antosha417** [reply]
> Or are you saying after you choose 1. bloop restarts, then you open it again it shows you it again?

Yes. If I open scala project and select `1.` then close neovim and open the same project again I see the message once again.

**Reply by ckipp01** [reply]
Huh, then yea that shouldn't be happening at all. Can you test something for me?

Do a

```
killall java
```
followed by a

```
jps -l
```
Make sure there is no Bloop instance running when you do the `jps`, and then start the project again. It might just be that something is going wrong with Metals and it's unable to shut down Bloop. If that was the case, then it'd detect the old Bloop every time and keep spamming you, which again, shouldn't happen.

**Reply by antosha417** [reply]
If I do `killall java` and then start the project again then there is no Bloop notification. And If I close neovim bloop is still running. So I think you are right and Metals is unable to shut down Bloop.

**Reply by ckipp01** [reply]
I think it might have just gotten stuck in a place where Bloop was possibly unresponsive and that happened to coincide with the update. If this happens again, comment back.

**Reply by antosha417** [reply]
It is **not** happening only if I kill bloop process by hand. I tried to reboot my laptop a couple of times and deleting .metals and .bloop directories. But it keeps happening.

**Reply by ckipp01** [reply]
What version of Metals and commit of `nvim-metals` are you on?

**Reply by antosha417** [reply]
`nvim-metals` 6738fe3c0e2142de01c753f8b7c18281d11488ce
`Metals` metals 0.11.6

**Reply by ckipp01** [reply]
One more thing, if you have Bloop cli installed can you paste the output of `bloop about`?

**Reply by ckipp01** [reply]
Actually nm, this morning I just hit on this exact same thing and opened an issue for it in https://github.com/scalameta/metals/issues/4001.

**Reply by ckipp01** [reply]
btw @antosha417 I pushed a fix for this and it's in the latest snapshot. If you set your `serverVersion` to `latest.snapshot` and then do a `:MetalsUpdate` you should get it. Can you verify that this fixes the issue for you as well?

**Reply by antosha417** [reply]
I don't see this issue anymore! Thank you a lot for such a quick fix! 
Always a pleasure to chat with you

**Reply by fiatjaf** [reply]
Should I set `serverVersion` to literally `"latest.snapshot"`?

**Reply by ckipp01** [reply]
`:h serverVersion` will show you this, yes.
